chia/util/files.py

```diff
@@ -0,0 +1,88 @@
+import asyncio
+import logging
+import os
+import shutil
+
+from aiofiles import tempfile  # type: ignore
+from pathlib import Path
+from typing import Union
+
+
+log = logging.getLogger(__name__)
+
+
+def move_file(src: Path, dst: Path):
+    """
+    Attempts to move the file at src to dst, falling back to a copy if the move fails.
+    """
+
+    dir_perms: int = 0o700
+    # Create the parent directory if necessary
+    os.makedirs(dst.parent, mode=dir_perms, exist_ok=True)
+
+    try:
+        # Attempt an atomic move first
+        os.replace(os.fspath(src), os.fspath(dst))
+    except Exception as e:
+        log.debug(f"Failed to move {src} to {dst} using os.replace, reattempting with shutil.move: {e}")
+        try:
+            # If that fails, use the more robust shutil.move(), though it may internally initiate a copy
+            shutil.move(os.fspath(src), os.fspath(dst))
+        except Exception:
+            log.exception(f"Failed to move {src} to {dst} using shutil.move")
+            raise
+
+
+async def move_file_async(src: Path, dst: Path, *, reattempts: int = 6, reattempt_delay: float = 0.5):
+    """
+    Attempts to move the file at src to dst, making multiple attempts if the move fails.
+    """
+
+    remaining_attempts: int = reattempts
+    while True:
+        try:
+            move_file(src, dst)
+        except Exception:
+            if remaining_attempts > 0:
+                log.debug(f"Failed to move {src} to {dst}, retrying in {reattempt_delay} seconds")
+                remaining_attempts -= 1
+                await asyncio.sleep(reattempt_delay)
+            else:
+                break
+        else:
+            break
+
+    if not dst.exists():
+        raise FileNotFoundError(f"Failed to move {src} to {dst}")
+    else:
+        log.debug(f"Moved {src} to {dst}")
+
+
+async def write_file_async(file_path: Path, data: Union[str, bytes], *, file_mode: int = 0o600):
+    """
+    Writes the provided data to a temporary file and then moves it to the final destination.
+    """
+
+    dir_perms: int = 0o700
+    # Create the parent directory if necessary
+    os.makedirs(file_path.parent, mode=dir_perms, exist_ok=True)
```

Comment:
is there a good reason to specify the permission bits here?
Do we specifically want to disallow the user from controlling this a more permissive `umask`?

Reply:
Having a more secure default seems like the right behavior for the case that the intermediates don't already exist. For use cases that want to ensure something more open, they can create the intermediates before calling this. We could have the directory permissions passed into the function.

Reply:
closing a file does not synchronize with the disk.

Reply:
ehm. what I mean is `close(fd)` doesn't block until the data has hit the disk. it's not trivial to find authoritative documentation on closing files, but flushing would have to be an `async` function, and there's no `await` call corresponding to closing of the context manager, is there?